In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.datasets import load_wine
from sklearn.model_selection import KFold
from sklearn.ensemble import BaggingClassifier, AdaBoostClassifier

In [3]:
#hago la carga del dataset con el q hay q trabajar
wine = load_wine()
X_wine = wine.data #caracteristicas del vino
y_wine = wine.target #clase a la que pertenece

In [4]:
#ahora vemos el dataset y si distribucion
print("Cantidad de patrones:", X_wine.shape[0])
print("Cantidad de características:", X_wine.shape[1])
print("Cantidad de clases:", len(np.unique(y_wine)))
print("Clases:", np.unique(y_wine))

Cantidad de patrones: 178
Cantidad de características: 13
Cantidad de clases: 3
Clases: [0 1 2]


vemos q hay 178 muestras descriptas por 13 caracteristicas y 3 clases de vino

In [5]:
#como en los ejs anteriores hago las particiones
kf = KFold(n_splits=5,shuffle=True,random_state=42)

In [6]:
#ahora creamos los modelso de ensamble con los q tenemos q trabajar
bagging = BaggingClassifier(random_state=42)

adaboost = AdaBoostClassifier(random_state=42)

In [7]:
#como en los ejs anteriores creo las listas de resultados
precision_bagging = []
precision_adaboost = []
#entrreno y evaluo
for fold, (train_idx, test_idx) in enumerate(kf.split(X_wine)):
    
    X_train = X_wine[train_idx]
    y_train = y_wine[train_idx]
    
    X_test = X_wine[test_idx]
    y_test = y_wine[test_idx]
    
    bagging.fit(X_train, y_train)
    accuracy_bagging = bagging.score(X_test, y_test)
    precision_bagging.append(accuracy_bagging)
    
    adaboost.fit(X_train, y_train)
    accuracy_adaboost = adaboost.score(X_test, y_test)
    precision_adaboost.append(accuracy_adaboost)
    
    print(f"Fold {fold + 1}:")
    print(f"  Bagging:  {accuracy_bagging:.4f}")
    print(f"  AdaBoost: {accuracy_adaboost:.4f}")

Fold 1:
  Bagging:  0.9444
  AdaBoost: 0.9444
Fold 2:
  Bagging:  1.0000
  AdaBoost: 0.9167
Fold 3:
  Bagging:  0.9722
  AdaBoost: 0.9722
Fold 4:
  Bagging:  0.9429
  AdaBoost: 0.9429
Fold 5:
  Bagging:  1.0000
  AdaBoost: 0.9714


lo q hacemos aca tiene la misma logica de antes, debemos recorrer cada fold y en cada uno entrenar y evaluar cada clasificador q nos pide
la consigna para luego plantear algun tipo de comparacion o sacar conclusiones.

In [9]:
media_bag=np.mean(precision_bagging)
var_bag=np.var(precision_bagging)

media_ada=np.mean(precision_adaboost)
var_ada=np.var(precision_adaboost)
print("Bagging:")
print(f"  Media: {media_bag:.4f}")
print(f"  Varianza: {var_bag:.6f}")

print("\nAdaBoost:")
print(f"  Media: {media_ada:.4f}")
print(f"  Varianza: {var_ada:.6f}")

Bagging:
  Media: 0.9719
  Varianza: 0.000635

AdaBoost:
  Media: 0.9495
  Varianza: 0.000429


podemos ver q el promedio bagging clasifico correctamente para este conjunto de datos un porcentaje mayor q el adaboost
aunque tambien podemos notar que adaboost posee una varianza menor lo q implica q los resultados fueron mas estables entre los distintos folds